### PR TITLE
Apple M-series GPU support

### PR DESCRIPTION
**Pull request by April-Hannah-Lena**
Very experimental. Just wanted to see if it would work. 

I was able to verify 100x performance improvement in a rough test while I was sitting in a coffee shop.

### CODE REVIEW

**Comment by April-Hannah-Lena** [comment]
Ok Apple M-series support is now working and passes tests. I also added some more tests for transfer operators in general. 

However, every once in a while a GPU computation causes a kernel panic, no idea why 😅😬 looking into it

**Comment by April-Hannah-Lena** [comment]
Yes exactly, metaltest.jl and the docs are next up to work on :)

**Comment by April-Hannah-Lena** [comment]
Ok, the docs are all updated and ready to merge 👍

File: docs/src/metal.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+    GAIO can convert your `BoxPartition` to 32-bit automatically when you use GPU acceleration, but preferred are still explicit 32-bit literals like
+    ```julia
+    center, radius = (0f0,0f0,25f0), (30f0,30f0,30f0)
+    ```
+    instead of `center, radius = (0,0,25), (30,30,30)`. 
+
+All we need to do is load the Metal.jl package and pass `:gpu` as the second argument to one of the two supported box map constructors: `BoxMap(:montecarlo, ...)`, `BoxMap(:grid, ...)`. Other box mapping algorithms are currently not supported. 
+```julia
+using CUDA
```
Comment (gaioguy) [inline docs/src/metal.md:36]: Why are you using `CUDA` here?

File: examples/modified_lorenz.jl [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+r = (25,25,25)
+domain = Box(c, r)
+P = BoxPartition(domain, (4,4,4))
+
+F = BoxMap(f, domain)
+A = cover(P, :)
+
+A = relative_attractor(F, A, steps=12)
+
+F = BoxMap(:gpusampled, f, domain, 750)
```
Comment (gaioguy) [inline examples/modified_lorenz.jl:20]: Why `:gpusampled` here instead of `:montecarlo, :gpu`?

File: examples/modified_lorenz.jl [hunk 3]
```diff
@@ -0,0 +1,39 @@
+
```
Comment (gaioguy) [inline examples/modified_lorenz.jl:1]: What is the purpose of this example?

File: src/GAIO.jl [hunk 4]
```diff
@@ -13,8 +13,18 @@ using FLoops
 using OrderedCollections
 using SplittablesBase
 using SplitOrderedCollections
+
+# interval arithmetic
 using IntervalArithmetic
 
+# newer version of IntervalArithmetic have changed many function names
```
Comment (gaioguy) [inline src/GAIO.jl:20]: Can we solve this more elegantly by requiring a certain package version at some place?

File: src/algorithms/invariant_sets.jl
Comment (gaioguy) [inline src/algorithms/invariant_sets.jl:1]: Can we consistently use `F#` for the transfer operator throughout all files?

File: src/algorithms/invariant_sets.jl [hunk 5]
```diff
@@ -26,7 +26,7 @@ e.g. `B = cover(P, :)` for a partition `P`.
 function chain_recurrent_set(F::BoxMap, B₀::BoxSet{Box{N,T}}; steps=12) where {N,T}
     B = copy(B₀)
     for k in 1:steps
-        B   = subdivide(B, (k % N) + 1)
+        B   = subdivide(B)
```
Comment (gaioguy) [inline src/algorithms/invariant_sets.jl:29]: Nice simplification!

File: src/algorithms/optimization.jl [hunk 6]
```diff
@@ -87,11 +87,10 @@ function cover_manifold(f, B₀::BoxSet{Box{N,T},Q,S}; steps=12) where {N,T,Q,S}
     for k in 1:steps
         B = subdivide(B, (k % N) + 1)
         P = B.partition
-        @floop for key in B.set
-            box = key_to_box(P, key)
-            int = IntervalBox(box)
-            fint = f(int)
-            if contains_zero(fint)
+        @floop for key in keys(B)
```
Comment (gaioguy) [inline src/algorithms/optimization.jl:90]: 👍

File: src/box.jl [hunk 7]
```diff
@@ -45,29 +45,18 @@ Box{T}(args...) where {T} = Box{T}(Box(args...))
...
-end
-
-function IntervalArithmetic.IntervalBox(box::Box{N,T}) where {N,T}
-    c, r = box
+Box{N,T}(ints::Interval...) where {N,T} = Box{N,T}(ints)
+Box{T}(ints::Interval...) where {T} = Box{T}(ints)
+Box(ints::SVNT{N,<:Interval{T}}) where {N,T} = Box{N,T}(ints)
+Box(ints::Interval...) = Box(ints)
+
+function Box{N,T}(ints::Container) where {N,T,Container<:SVNT{N,Interval{T}}}
```
Comment (gaioguy) [inline src/box.jl:53]: 👍

File: src/boxmap_intervals.jl
Comment (gaioguy) [inline src/boxmap_intervals.jl:1]: Again a nice simplification 👍

File: docs/src/metal.md [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+center, radius = (0f0,0f0,25f0), (30f0,30f0,30f0)
+Q = Box(center, radius)
+P = BoxPartition(Q, (128,128,128))
+F = BoxMap(:montecarlo, :gpu, f, Q)
+
+x = (sqrt(β*(ρ-1)), sqrt(β*(ρ-1)), ρ-1)
+S = cover(P, x)
+@time W = unstable_set(F, S)
+```
```
Comment (gaioguy) [inline docs/src/metal.md:46]: When I run this, I get the error
```
julia> F = BoxMap(:montecarlo, :gpu, f, Q)
ERROR: MethodError: no method matching MetalExt.GPUSampledBoxMap(::typeof(f), ::Box{3, Float64}, ::MtlVector{StaticArraysCore.SVector{3, Float32}, Private})

Closest candidates are:
  MetalExt.GPUSampledBoxMap(::F, ::Box{N, T}, ::MtlArray{StaticArraysCore.SVector{N, T}}) where {N, T, F<:Function}
   @ MetalExt ~/Applications/April/GAIO.jl/ext/MetalExt.jl:24

Stacktrace:
 [1] MonteCarloBoxMap(::Val{:gpu}, map::Function, domain::Box{3, Float64}; n_points::Int64)
   @ MetalExt ~/Applications/April/GAIO.jl/ext/MetalExt.jl:180
 [2] MonteCarloBoxMap(::Val{:gpu}, map::Function, domain::Box{3, Float64})
   @ MetalExt ~/Applications/April/GAIO.jl/ext/MetalExt.jl:178
 [3] BoxMap(::Val{:montecarlo}, ::Val{:gpu}, ::Vararg{Any}; kwargs::@Kwargs{})
   @ GAIO ~/Applications/April/GAIO.jl/src/boxmap.jl:28
 [4] BoxMap(::Symbol, ::Val{:gpu}, ::Vararg{Any}; kwargs::@Kwargs{})
   @ GAIO ~/Applications/April/GAIO.jl/src/boxmap.jl:12
 [5] BoxMap(::Symbol, ::Symbol, ::Function, ::Vararg{Any}; kwargs::@Kwargs{})
   @ GAIO ~/Applications/April/GAIO.jl/src/boxmap.jl:13
 [6] top-level scope
   @ REPL[14]:1
 [7] top-level scope
   @ ~/.julia/packages/Metal/TVRIh/src/initialization.jl:64
```

**Review by gaioguy — COMMENTED** [review]
Great work! I have a few minor questions and comments - and the example given in docs does not run yet ...

**Pull request merged**